Feature/reg state dao fix

seoklee opened the pull request:
- Removed HashMap and used DB instead for registrationState. 
- .gitignore to newly added modules

waynemz commented on hni-events/src/main/java/org/hni/events/service/AbstractRegistrationService.java:
```diff
@@ -26,10 +26,10 @@
         if (state == null) {
             // No state, so we're just beginning.
-            state = new RegistrationState(EventName.REGISTER, event.getPhoneNumber(),
+            state = new RegistrationState(null, EventName.REGISTER, event.getPhoneNumber(),
```
setting the id to any value doesn't do anything because this object is 'detached'. It get attached only after doing an insert below. You could use a constructor that doesn't take the id. Instead of a constructor, better use a factory method to create detached registrationState objects. having a no-argument constructor implies that you can create an object in fully initialized form but that is not the case here.

seoklee replied:
I will make a factory method that will not accept the id as the parameter.

waynemz commented on hni-events/src/main/java/org/hni/events/service/AbstractRegistrationService.java:
```diff
@@ -41,9 +41,8 @@ public String handleEvent(final Event event) {
         if (!state.getRegistrationStep().equals(stepResult.getNextStateCode())) {
             final RegistrationState nextState =
-                    new RegistrationState(state.getEventName(), state.getPhoneNumber(),
+                    new RegistrationState(state.getId(), state.getEventName(), state.getPhoneNumber(),
```
you are creating a new detached entity here with the updated state and trying to persist it later. You should always try to update attached enitity object instead.

waynemz replied:
The detached object is not managed by the entity manager and does not represent a row in the table unless it is merged. It generally doesn't make sense to create an immutable entity if you are using JPA and expecting to update the entity. Then by definition the entity object must not be mutable.

seoklee replied:
I see. I will make the change.

waynemz commented on hni-events/src/main/java/org/hni/events/service/om/RegistrationState.java:
```diff
@@ -22,30 +22,24 @@
-    }
-
-    public RegistrationState(EventName eventName, String phoneNumber, String payload, RegistrationStep registrationStep) {
+    public RegistrationState(Long id, EventName eventName, String phoneNumber, String payload, RegistrationStep registrationStep) {
```
see my comment about the factory method above

seoklee commented:
@waynemz I added setters for RegistrationState and additional factory method to create a detached entity. 

Since RegistrationState is no longer immutable object, I removed a WorkFlowResult as a result.

waynemz commented on hni-events/src/test/java/org/hni/events/service/RegisterServiceUnitTest.java:
```diff
@@ -49,97 +49,89 @@ public void setUp() {
     @Test
     public void testGetFirstName() throws Exception {
-        state = new RegistrationState(REGISTRATION_STATE_ID, EventName.REGISTER, PHONE_NUMBER, payload, RegistrationStep.STATE_REGISTER_GET_FIRST_NAME);
+        state = RegistrationState.create(EventName.REGISTER, PHONE_NUMBER, payload, RegistrationStep.STATE_REGISTER_GET_FIRST_NAME);
```
what happened to the REGISTRATION_STATE_ID constant here? We don't need it?

seoklee replied:
REGISTRATION_STATE_ID constant was there to simulate id given from JPA. However, we removed the id parameter in this particular constructor, and created a new factory method instead because of our previous conversation.

waynemz commented on hni-events/src/main/java/org/hni/events/service/AbstractRegistrationService.java:
```diff
@@ -16,7 +16,7 @@
  * Handles Registration events.
  */
-public abstract class AbstractRegistrationService<T> implements EventService{
+public abstract class AbstractRegistrationService<T> implements EventService {
```
At this point there isn't much left in this abstract class. you could combine this code into RegisterService

seoklee replied:
I think that'd be more fitting in the other refactoring PR that I will publish later.

waynemz commented:
@seoklee were you able to test these changes on mysql? do they work?

waynemz commented:
Holding this due to the fact this is a big change and requires more local testing against the test data set to assure confidence.

seoklee commented:
Did some local testing, and registration seems to work fine as it is.

seoklee commented:
issue 119
